sympy/geometry/line.py

```diff
@@ -1280,6 +1280,57 @@ def distance(self, other):
             return S.Zero
         return self.perpendicular_segment(other).length
 
+    def nearest_points(l1, l2):
+        """
+        return the closest distance of separation between l1 and l2
+        and the points connecting the lines with that distance of
+        separation.
+
+        Examples
+        ========
+
+        >>> from sympy import Line, Point
+        >>> f = Line.nearest_points
+        >>> l3 = Line(Point(0, 0, 0), Point(-1, 1, 0))
+
+        There is no separation between collinear lines:
+
+        >>> l3b = Line(Point(0, 0, 0), Point(1, -1, 0))
+        >>> f(l3, l3b)
+        (0, Point3D(-t, t, 0))
+
+        There is a uniform distance between non-intersecting
+        parallel lines:
+
+        >>> lp = Line(Point(1, 0, 0), Point(0, 1, 0))
+        >>> f(l3, lp)
+        (sqrt(2)/2, [Point3D(-t, t, 0), Point3D(1/2 - t, t + 1/2, 0)])
+        >>> l4 = Line(Point(1, 0, 1), Point(1, 1, 1))
+        >>> f(l3, l4)
+        (1, [Point3D(1, -1, 0), Point3D(1, -1, 1)])
+        """
+        from sympy.core.symbol import uniquely_named_symbol
+        from sympy.solvers.solvers import solve
+        u = uniquely_named_symbol('t', (l1, l2))
+        same = l1.equals(l2)
+        if same:
+            return S.Zero, l1.arbitrary_point(u)
+        assert same == False  # if we aren't sure, don't continue
+        p1 = l1.arbitrary_point(u)
+        if l1.is_parallel(l2):
+            d = l2.distance(p1)
+            p2 = l1.perpendicular_line(p1).intersection(l2)[0]
+            return d, [p1, p2]
+        p2 = l2.arbitrary_point(u)
+        pts = []
+        for i in range(2):
+            d = l2.distance(p1)
+            v = solve(d.diff(u), u)[0]
+            pts.append(p1.subs(u, v))
+            l2, l1 = l1, l2
+            p1, p2 = p2, p1
+        return d.subs(u, v), pts
```

Comment:
Is it similar to https://brilliant.org/wiki/3d-coordinate-geometry-distance/#distance-between-2-skew-lines or https://math.stackexchange.com/a/2217845? Distance will only matter if the lines are non-parallel, the formula in the above links should work in that case. Not sure though.

Reply:
Yes @smichr `close_points ` answers are not consistent throughout .I've been a bit dicey after some of it's answers wouldn't match with `nearest_points` earlier .
This is one I remember right now but I have encountered more ( this is a parallel case ,I believe it will be also incorrect in collinear cases)
```
>>> l1=Line(Point(0,0,0), Point(-1,1,0))
>>> l2= Line(Point(1, 0, 0), Point(0, 1, 0))
>>> separation(l1,l2)
nan

>>> close_points(l1,l2)
Traceback (most recent call last):
  File "C:\Users\kunni\sympy\sympy\sympy\geometry\point.py", line 218, in __add__
    s, o = Point._normalize_dimension(self, Point(other, evaluate=False))
  File "C:\Users\kunni\sympy\sympy\sympy\geometry\point.py", line 125, in __new__
    raise TypeError(filldedent('''
TypeError:
Expecting sequence of coordinates, not `NaN`
```